### PR TITLE
Hold logs as JSON

### PR DESCRIPTION
**Pull request by nemaniarjun**
When `coala --json` is run, user should have a option
to get the logs in form of JSON

Closes https://github.com/coala/coala/issues/3037

### Checklist

- [ ] I have rebased properly. Please see [our tutorial on rebasing](http://coala.io/git#rebasing).
- [ ] I have gone through the [commit guidelines](http://coala.io/commit) and I've followed them.
- [ ] I have followed the [guidelines for the commit shortlog](http://coala.io/commit#shortlog).
- [ ] I have followed the [guidelines for the commit body](http://coala.io/commit#commit-body).
- [ ] I have included the issue URL with the 'Fixes' keyword if the commit fixes a *real bug* and the 'Closes' keyword if it adds a feature or enhancement. For details, check the [issue reference section in our commit guidelines](http://coala.io/commit#issue-reference).
- [ ] I have [run all the tests](http://api.coala.io/en/latest/Developers/Executing_Tests.html) and they all pass. If any CI (below) is not green, please fix them. If GitMate issues appear you will have to amend your commits, pushing new commits on top is not sufficient!

### CODE REVIEW

**Comment by nemaniarjun** [comment]
CC @sils @makmam2

File: coalib/coala_main.py [hunk 1]
```diff
@@ -63,7 +62,6 @@ def run_coala(console_printer=None,
     :return:                        A dictionary containing a list of results
                                     for all analyzed sections as key.
     """
-    configure_logging()
```
Comment (nemaniarjun) [inline coalib/coala_main.py]: This was redundent. We are already configuring logging in `coala.py`

**Comment by sils** [comment]
CC @tulphoon he's working on some other logging stuff and might have valuable input. Don't cc me everywhere please :(

File: coalib/misc/Constants.py [hunk 2]
```diff
@@ -128,39 +128,3 @@
...
-                'datefmt': '%X',
-                'log_colors': {
-                    'ERROR': 'red',
-                    'WARNING': 'yellow',
-                    'INFO': 'blue',
-                    'DEBUG': 'green'
-                }
-            }
-        }
-    })
```
Comment (sils) [inline coalib/misc/Constants.py]: can you put the move of this function into an own commit? This just clutters the review. Do a `git reset HEAD~1` and then `git add -p` add onlyt the changes you want to commit and `git commit` (without -a!) it, then commit the rest :)

**Comment by sils** [comment]
approach looks good!

File: coalib/misc/Constants.py (hunk 2 above)
Comment (nemaniarjun) [inline coalib/misc/Constants.py]: Okay. Will do that. Also I think I wont need git guidelines anymore. :wink:

Comment (sils) [inline coalib/misc/Constants.py]: sorry, most people don't think of reset and add -p I think it's not in any of our tutorials even so I figured it might save you some time to mention it ;)

Comment (nemaniarjun) [inline coalib/misc/Constants.py]: Should we add it?
Or is it too advanced and might frighten the newbies?

File: coalib/output/Logging.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        'root': {
+            'level': 'DEBUG',
+            'handlers': ['json']
+        },
+        'formatters': {
+            'json': {
+                '()': 'logging.Formatter',
+                'format': '{ "message": "%(message)s",'
+                          ' "level": "%(levelname)s",'
+                          ' "time": "%(asctime)s" }|',
```
Comment (nemaniarjun) [inline coalib/output/Logging.py]: Note that I am adding `|` at end of each log. Then `str.split`ing them. This means if there is a `|` in any log message, this might break.
Need suggestions on how to implement it better. (Maybe some char thats never used anywhere?)

Comment (Makman2) [inline coalib/output/Logging.py]: can't you avoid this? I mean `|` is on my keyboard, this case can happen^^

File: coalib/output/Logging.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+
+def parse_json_log(stream):
+    """
+    Parses A Stream of JSON Handler
+    Returns a list of JSON object for logs
+    """
+    import json
+
+    logs = list(map(str.strip, stream.getvalue().split('|')))
```
Comment (Makman2) [inline coalib/output/Logging.py]: you don't need to collect in list again, keep it as an iterator as you use it once below^^

File: coalib/output/Logging.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    logs = [json.loads(log, strict=False) for log in logs if log]
+
+    return logs
```
Comment (Makman2) [inline coalib/output/Logging.py]: if I understand correctly you use this function to grab our complete logging output and transform that into multiple log-objects in JSON. Can't you use acess the logs in `logging` as a list/collection or so, or use another logger where you override the log-method and put each log-message into a list (maybe something like this does already exist)?

Comment (sils) [inline coalib/output/Logging.py]: yep, there *must* be a better way, writing a custom formatter that gets the actual log objects with the data instead of formatting it and reparsing it back

File: coalib/output/Logging.py (hunk 3 above)
Comment (nemaniarjun) [inline coalib/output/Logging.py]: Okay! Any suggestion for some other character?

File: coalib/output/Logging.py (hunk 4 above)
Comment (nemaniarjun) [inline coalib/output/Logging.py]: :+1:

File: coalib/output/Logging.py (hunk 5 above)
Comment (nemaniarjun) [inline coalib/output/Logging.py]: Will look into this. As I mentioned in the issue, [this](https://github.com/madzak/python-json-logger) does what we want. But it's not well documented. Will try to configure it out again today.

Comment (nemaniarjun) [inline coalib/output/Logging.py]: Do we need the time of log? Since this option is going to be used after coala has finished running. Is time really an important factor to the log?

Comment (Makman2) [inline coalib/output/Logging.py]: I think time is not bad and is an essential part of a log message. Why it's a problem to get the time?

Comment (nemaniarjun) [inline coalib/output/Logging.py]: No, its not.

Comment (nemaniarjun) [inline coalib/output/Logging.py]: I have implemented a new method

File: coalib/output/Logging.py [hunk 6]
```diff
@@ -33,3 +36,64 @@ def configure_logging():
...
+    """
+
+    def __init__(self):
+        pass
+
+    @classmethod
+    def format_timestamp(cls, time):
+        from datetime import datetime
+
+        return datetime.utcfromtimestamp(time).isoformat() + 'Z'
```
Comment (Makman2) [inline coalib/output/Logging.py]: `Z` ? :)

File: coalib/output/Logging.py [hunk 7]
```diff
@@ -33,3 +36,64 @@ def configure_logging():
...
+    """
+    JSON formatter for python logging
+    """
+
+    def __init__(self):
+        pass
+
+    @classmethod
+    def format_timestamp(cls, time):
+        from datetime import datetime
```
Comment (Makman2) [inline coalib/output/Logging.py]: I would import at the top of the module^^

File: coalib/output/Logging.py [hunk 8]
```diff
@@ -33,3 +36,64 @@ def configure_logging():
...
+
+class JSONFormatters(logging.Formatter):
+    """
+    JSON formatter for python logging
+    """
+
+    def __init__(self):
+        pass
+
+    @classmethod
```
Comment (Makman2) [inline coalib/output/Logging.py]: can be made a staticmethod, you don't use `cls`

File: coalib/output/Logging.py [hunk 9]
```diff
@@ -33,3 +36,64 @@ def configure_logging():
...
+    def format_timestamp(cls, time):
+        from datetime import datetime
+
+        return datetime.utcfromtimestamp(time).isoformat() + 'Z'
+
+    def format(self, record, serialize=True):
+        import json
```
Comment (Makman2) [inline coalib/output/Logging.py]: import at top :)

File: coalib/output/Logging.py [hunk 10]
```diff
@@ -33,3 +36,64 @@ def configure_logging():
...
+    def __init__(self):
+        pass
+
+    @classmethod
+    def format_timestamp(cls, time):
+        from datetime import datetime
+
+        return datetime.utcfromtimestamp(time).isoformat() + 'Z'
+
+    def format(self, record, serialize=True):
```
Comment (Makman2) [inline coalib/output/Logging.py]: can be made a staticmethod or classmethod, you don't need an instance to access `format`

File: coalib/output/Logging.py [hunk 11]
```diff
@@ -33,3 +36,64 @@ def configure_logging():
...
+    return stream
+
+
+class JSONFormatters(logging.Formatter):
+    """
+    JSON formatter for python logging
+    """
+
+    def __init__(self):
+        pass
```
Comment (Makman2) [inline coalib/output/Logging.py]: do you need this one?

File: coalib/output/Logging.py (hunk 10 above)
Comment (nemaniarjun) [inline coalib/output/Logging.py]: I dont know enough about classmethod or staticmethod to understand this comment. The other changes have been made. :)

Comment (Makman2) [inline coalib/output/Logging.py]: then it's time to engage, and read up on google / python docs :)

Comment (nemaniarjun) [inline coalib/output/Logging.py]: I did understand what classmethod and staticmethod mean, but why/how do they apply here.

Comment (nemaniarjun) [inline coalib/output/Logging.py]: Heres what I understood. A staticmethod function is some function that does not require `self` (instance of an class). A classmethod function is some function that requires `cls` (the class itself).

Here since we dont need `cls` as we are not using any other functions/variables that are defined in the class or its parent classes we wont use classmethod funtion.

And since we also dont need instance functions/variables we will use a staticmethod funtion.

File: coalib/output/Logging.py [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+    })
+    return stream
+
+
+class JSONFormatters(logging.Formatter):
+    """
+    JSON formatter for python logging
+    """
+
+    def format(record):
```
Comment (Makman2) [inline coalib/output/Logging.py]: you have to decorate it with `@staticmethod`, otherwise your function won't work and `record` is going to be filled with the `self` instance^^

File: coalib/output/Logging.py [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+    Returns a StringIO() that captures the logs as json.
+    """
+    import logging.config
+    import io
+
+    stream = io.StringIO()
+
+    logging.config.dictConfig({
+        'version': 1,
+        'handlers': {
```
Comment (nemaniarjun) [inline coalib/output/Logging.py]: Isnt this a falst alarm?

Comment (Makman2) [inline coalib/output/Logging.py]: looks like

**Comment by nemaniarjun** [comment]
Waiting for more comments.

File: coalib/coala_delete_orig.py [hunk 14]
```diff
@@ -2,12 +2,12 @@
 
 from pyprint.ConsolePrinter import ConsolePrinter
 
+from coalib.output.Logging import configure_logging
 from coalib.output.printers.LogPrinter import LogPrinter
 from coalib.parsing import Globbing
 from coalib.settings.ConfigurationGathering import get_config_directory
 from coalib.settings.Section import Section
 from coalib.parsing.Globbing import glob_escape
-from coalib.misc.Constants import configure_logging
```
Comment (Makman2) [inline coalib/coala_delete_orig.py]: unrelated change you should move this into an own commit :)

File: coalib/coala_main.py (hunk 1 above)
Comment (Makman2) [inline coalib/coala_main.py]: I think this one could also go into a separate commit, there you can also put your explanation into the commit body :+1:

File: coalib/coala_modes.py [hunk 15]
```diff
@@ -66,6 +70,20 @@ def mode_json(args):
         results, exitcode, _ = run_coala()
 
     retval = {'bears': results} if args.show_bears else {'results': results}
+
+    if args.log_json:
+        logs = []
+        for line in log_stream.getvalue().splitlines():
+            logs.append(json.loads(line))
+        retval = {'logs': logs}
```
Comment (Makman2) [inline coalib/coala_modes.py]: you override `retval` again, this does not look right^^ If you want to update `retval`, just do `retval['logs'] = logs`

File: coalib/coala_modes.py [hunk 16]
```diff
@@ -66,6 +70,20 @@ def mode_json(args):
...
+    else:
+        retval = {}
+
+    if args.show_bears:
+        retval['bears'] = results
+    else:
+        retval['results'] = results
```
Comment (Makman2) [inline coalib/coala_modes.py]: -->
```
retval['bears' if args.show_bears else 'results'] = results
```

**Comment by nemaniarjun** [comment]
I think this is better. @Makman2 changes made.

File: coalib/coala_delete_orig.py (hunk 14 above)
Comment (Makman2) [inline coalib/coala_delete_orig.py]: oh indeed, nvm ... :3

File: coalib/coala_main.py [hunk 17]
```diff
@@ -13,7 +13,6 @@
 from coalib.misc.Caching import FileCache
```
Comment (Makman2) [inline coalib/coala_main.py]: commit message:
it's "redund**a**nt" :)

File: coalib/output/Logging.py [hunk 18]
```diff
@@ -33,3 +38,53 @@ def configure_logging():
             }
         }
     })
+
+
+def configure_json_logging():
+    """
+    Configures logging as a JSON objects.
+    Returns a StringIO() that captures the logs as json.
```
Comment (Makman2) [inline coalib/output/Logging.py]: 1. --> `Configures logging for JSON.`
2. You should use `:return:`
   ```
   :return: Returns a ``StringIO`` that captures the logs as JSON.
   ```

File: coalib/output/Logging.py [hunk 19]
```diff
@@ -33,3 +38,53 @@ def configure_logging():
...
+    """
+    import logging.config
+    import io
```
Comment (Makman2) [inline coalib/output/Logging.py]: why not importing them at the top of the module?

File: coalib/output/Logging.py [hunk 20]
```diff
@@ -33,3 +38,53 @@ def configure_logging():
...
+        'formatters': {
+            'json': {
+                '()': 'coalib.output.Logging.JSONFormatters',
+            }
+        }
+    })
+    return stream
+
+
+class JSONFormatters(logging.Formatter):
```
Comment (Makman2) [inline coalib/output/Logging.py]: please just `JSONFormatter` :+1:

File: coalib/output/Logging.py [hunk 21]
```diff
@@ -33,3 +38,53 @@ def configure_logging():
...
+    """
+    @staticmethod
+    def format(record):
+        # Create message dict
+        message = {
+            'timestamp': datetime.utcfromtimestamp(record.created).isoformat(),
+            'message': record.getMessage(),
+            # 'path': record.pathname,
+            'level': record.levelname,
+            # 'logger': record.name,
```
Comment (Makman2) [inline coalib/output/Logging.py]: please remove commented code :+1:

File: coalib/output/Logging.py [hunk 22]
```diff
@@ -33,3 +38,53 @@ def configure_logging():
...
+    return stream
+
+
+class JSONFormatters(logging.Formatter):
+    """
+    JSON formatter for python logging
+    """
+    @staticmethod
+    def format(record):
+        # Create message dict
```
Comment (Makman2) [inline coalib/output/Logging.py]: I think it's understandable what the below dict is for ;)

File: coalib/output/Logging.py [hunk 23]
```diff
@@ -33,3 +38,53 @@ def configure_logging():
...
+                '()': 'coalib.output.Logging.JSONFormatters',
+            }
+        }
+    })
+    return stream
+
+
+class JSONFormatters(logging.Formatter):
+    """
+    JSON formatter for python logging
```
Comment (Makman2) [inline coalib/output/Logging.py]: Plase add a period at the end of the sentence ;)

File: coalib/coala_modes.py [hunk 24]
```diff
@@ -48,6 +48,10 @@ def mode_json(args):
     from coalib.output.printers.LogPrinter import LogPrinter
     from coalib.settings.ConfigurationGathering import get_filtered_bears
 
+    if args.log_json:
+        from coalib.output.Logging import configure_json_logging
```
Comment (Makman2) [inline coalib/coala_modes.py]: why not importing at top of module?

**Comment by Makman2** [comment]
ack 644908d

**Comment by Makman2** [comment]
unack a673a6e, 3a0e275

File: coalib/coala_modes.py [hunk 25]
```diff
@@ -48,6 +48,10 @@ def mode_json(args):
     from coalib.output.printers.LogPrinter import LogPrinter
```
Comment (Makman2) [inline coalib/coala_modes.py]: commit message:
I think you should point out that you add a `--log-json` flag^^
sth like: `Add --log-json flag` or so

File: coalib/output/Logging.py (hunk 21 above)
Comment (nemaniarjun) [inline coalib/output/Logging.py]: I kept it for future reference. Do we need the path of the file in which the error originated? 
Also if we add multiple loggers in the future (we use only one now) we might need the logger key.

File: coalib/output/Logging.py (hunk 19 above)
Comment (nemaniarjun) [inline coalib/output/Logging.py]: I usually import only when things are needed. Is that not what I should do?

Comment (Makman2) [inline coalib/output/Logging.py]: no it's just we usually import everything at the top of the module. Especially in your case I'm not sure what's better, as your function will be normally only called once in an application lifecycle^^

So: keep it :+1: ;)

File: coalib/output/Logging.py (hunk 21 above)
Comment (Makman2) [inline coalib/output/Logging.py]: commented out code is bad, it does nothing and does not necessary help, but adding more text into the file ;)

> Do we need the path of the file in which the error originated?

We don't print to stdout, so currently it's not necessary^^

> Also if we add multiple loggers in the future (we use only one now) we might need the logger key.

We don't use multiple loggers yet, so no need for this ;)

**Comment by nemaniarjun** [comment]
@Makman2 Addressed All comments

**Comment by Makman2** [comment]
Ah sorry I meant the shortlog :3

> commit message:
> I think you should point out that you add a --log-json flag^^
> sth like: Add --log-json flag or so

For `JSON Mode: Hold logs as JSON`

**Comment by nemaniarjun** [comment]
And.. Done.

**Comment by nemaniarjun** [comment]
I am not too confident about the tests. Can you look at them once again?

File: coalib/coala_modes.py [hunk 26]
```diff
@@ -66,6 +70,11 @@ def mode_json(args):
         results, exitcode, _ = run_coala()
 
     retval = {'bears': results} if args.show_bears else {'results': results}
+
+    if args.log_json:
+        retval['logs'] = [json.loads(line) for line in
+                          log_stream.getvalue().splitlines()]
```
Comment (Makman2) [inline coalib/coala_modes.py]: why not `retval['logs'] = log_stream.getvalue()`?

Comment (Makman2) [inline coalib/coala_modes.py]: nvm

**Comment by nemaniarjun** [comment]
Tests fail due to some unrelated error.

**Comment by Makman2** [comment]
@nemaniarjun please do a `git commit --amend --no-edit; git push -f` next time to avoid having the close inside the history ;)

File: tests/output/printers/LoggingTest.py [hunk 27]
```diff
@@ -0,0 +1,33 @@
...
+        logs_list = []
+        for line in log_stream.getvalue().splitlines():
+            logs_list.append(json.loads(line))
+
+        self.assertEqual(logs_list[0]['message'], 'This is debug log.')
+        self.assertEqual(logs_list[0]['level'], 'DEBUG')
+        self.assertEqual(logs_list[1]['message'], 'This is info log.')
+        self.assertEqual(logs_list[1]['level'], 'INFO')
+        self.assertEqual(logs_list[2]['message'], 'This is warning log.\n'
+                         ' This is continued')
```
Comment (Makman2) [inline tests/output/printers/LoggingTest.py]: I think this looks more beautiful and is better to read ;)
```
self.assertEqual(logs_list[2]['message'],
                 'This is warning log.\n This is continued')
```

File: tests/output/printers/LoggingTest.py [hunk 28]
```diff
@@ -0,0 +1,33 @@
...
+        log_stream = configure_json_logging()
+
+        logging.debug('This is debug log.')
+        logging.info('This is info log.')
+        logging.warning('This is warning log.\n This is continued')
+        logging.error('This is error log.')
+        # This is a list of logs dicts.
+        # Each log dict has a `message`, `level` and `time`
+        # We will check message and level for each of the logs
+
```
Comment (Makman2) [inline tests/output/printers/LoggingTest.py]: though it's just a comment, you should move the newline above it, as it addresses the `logs_list` variable ;)

File: tests/output/printers/LoggingTest.py [hunk 29]
```diff
@@ -0,0 +1,33 @@
+import unittest
+import logging
+import json
```
Comment (Makman2) [inline tests/output/printers/LoggingTest.py]: though we don't have the bear ready yet, could you sort those imports in alphabetical order?

File: coalib/output/Logging.py [hunk 30]
```diff
@@ -1,3 +1,10 @@
+from datetime import datetime
+import json
+import logging
+import logging.config
+import io
```
Comment (Makman2) [inline coalib/output/Logging.py]: if possible sort these alphabetically ;)

File: tests/output/printers/LoggingTest.py (hunk 29 above)
Comment (nemaniarjun) [inline tests/output/printers/LoggingTest.py]: bear..? :P

**Comment by nemaniarjun** [comment]
image

File: tests/output/LoggingTest.py [hunk 31]
```diff
@@ -0,0 +1,33 @@
...
+        logging.info('This is info log.')
+        logging.warning('This is warning log.\n This is continued')
+        logging.error('This is error log.')
+
+        # This is a list of logs dicts.
+        # Each log dict has a `message`, `level` and `time`
+        # We will check message and level for each of the logs
+        logs_list = []
+        for line in log_stream.getvalue().splitlines():
+            logs_list.append(json.loads(line))
```
Comment (Makman2) [inline tests/output/LoggingTest.py]: you could use a generator expression:
```
logs_list = [json.loads(line)
             for line in log_stream.getvalue().splitlines()]
```

**Comment by nemaniarjun** [comment]
LGTM! @Makman2

**Comment by Makman2** [comment]
ack 944ce16, 6680ded, 1724844

**Comment by Makman2** [comment]
Cherry picked into https://github.com/coala/coala/pull/3594
